File: examples/python/m2-tool-runtime/one_step.py
"""Fake Transport 教学用的固定一步闭环；它不是 Agent Loop。

M1 不保留 reasoning 等未知输出项，因此本模块不应直接用于生产级真实多轮
Responses 调用。完整上下文保留或 previous_response_id 将在 M3 设计。
"""

from __future__ import annotations

from dataclasses import dataclass
import sys
from pathlib import Path

from bridge import result_to_message, spec_to_tool_definition, tool_use_to_call
from registry import ToolRegistry
from tool_types import ToolResult

_M1 = Path(__file__).resolve().parents[1] / "m1-unified-protocol"
if str(_M1) not in sys.path:
    sys.path.insert(0, str(_M1))

from chat_once import Config, HttpTransport, complete  # noqa: E402
from protocol import ModelRequest, ModelResponse  # noqa: E402


class OneStepError(RuntimeError):
    """固定闭环的形状不符合“恰好一次工具调用”时抛出。"""


@dataclass(frozen=True)
class OneStepResult:
    """保留闭环的关键证据，便于测试 call_id 和第二次请求。"""

    first_response: ModelResponse
    tool_result: ToolResult
    followup_request: ModelRequest
    final_response: ModelResponse


def request_with_registry_tools(
    request: ModelRequest, registry: ToolRegistry
) -> ModelRequest:
    """从 Registry 生成有序的模型工具声明，并复制其余统一请求字段。"""

    return ModelRequest(
        model=request.model,
        system=request.system,
        messages=request.messages,
        tools=tuple(spec_to_tool_definition(spec) for spec in registry.specs()),
    )


# ANCHOR: m2-one-step
def run_one_tool_step(
    request: ModelRequest,
    config: Config,
    transport: HttpTransport,
    registry: ToolRegistry,
) -> OneStepResult:
    """用可控教学传输调用两次，中间只允许并执行恰好一个工具候选。

    第一次没有工具、返回多个工具，或第二次再次要工具都会立即失败；本函数
    不重试、不循环，也不定义预算和停止策略。生产级真实响应还必须保留 M1
    未建模的输出项，本函数刻意不承担该职责。
    """

    first_request = request_with_registry_tools(request, registry)
    first = complete(first_request, config, transport)
    tool_uses = first.tool_uses()
    if len(tool_uses) != 1:
        raise OneStepError(
            f"first model call must request exactly one tool; got {len(tool_uses)}"
        )

    result = registry.execute(tool_use_to_call(tool_uses[0]))
    followup = ModelRequest(
        model=first_request.model,
        system=first_request.system,
        messages=first_request.messages + (first.message, result_to_message(result)),
        tools=first_request.tools,
    )
    final = complete(followup, config, transport)
    if final.tool_uses():
        raise OneStepError("second model call requested another tool; Agent Loop is M3")
    if not final.text():
        raise OneStepError("second model call returned no final text")
    return OneStepResult(first, result, followup, final)
# ANCHOR_END: m2-one-step

File: examples/rust/m2-tool-runtime/src/one_step.rs
//! Fake Transport 教学用的固定一步闭环；它不是 Agent Loop。
//!
//! M1 不保留 reasoning 等未知输出项，因此这里不用于生产级真实多轮调用。

use crate::bridge::{result_to_message, spec_to_tool_definition, tool_use_to_call};
use crate::{ToolError, ToolRegistry, ToolResult};
use m1_unified_protocol::{Config, HttpTransport, ModelRequest, ModelResponse, complete};
use std::fmt;

/// 固定闭环形状不满足约束时的错误。
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OneStepError(pub String);

impl fmt::Display for OneStepError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for OneStepError {}

impl From<ToolError> for OneStepError {
    fn from(error: ToolError) -> Self {
        Self(error.to_string())
    }
}

/// 保留闭环关键证据，供测试检查 call_id 与第二次请求。
#[derive(Debug)]
pub struct OneStepResult {
    pub first_response: ModelResponse,
    pub tool_result: ToolResult,
    pub followup_request: ModelRequest,
    pub final_response: ModelResponse,
}

/// 从 Registry 生成有序模型工具声明，并复制请求其余字段。
pub fn request_with_registry_tools(
    request: &ModelRequest,
    registry: &ToolRegistry,
) -> Result<ModelRequest, OneStepError> {
    let tools = registry
        .specs()
        .iter()
        .map(spec_to_tool_definition)
        .collect::<Result<Vec<_>, _>>()?;
    ModelRequest::try_new_with_tools(
        request.model.clone(),
        request.messages.clone(),
        request.system.clone(),
        tools,
    )
    .map_err(|error| OneStepError(error.to_string()))
}

/// 第一次恰好一个工具候选，执行后第二次必须给出最终文本。
pub fn run_one_tool_step<T: HttpTransport>(
    request: &ModelRequest,
    config: &Config,
    transport: &mut T,
    registry: &mut ToolRegistry,
) -> Result<OneStepResult, OneStepError> {
    let first_request = request_with_registry_tools(request, registry)?;
    let first = complete(&first_request, config, transport)
        .map_err(|error| OneStepError(error.to_string()))?;
    let uses = first.tool_uses();
    if uses.len() != 1 {
        return Err(OneStepError(format!(
            "first model call must request exactly one tool; got {}",
            uses.len()
        )));
    }
    let call = tool_use_to_call(uses[0]);
    let result = registry.execute(call);
    let mut messages = first_request.messages.clone();
    messages.push(first.message.clone());
    messages.push(result_to_message(&result)?);
    let followup = ModelRequest::try_new_with_tools(
        first_request.model.clone(),
        messages,
        first_request.system.clone(),
        first_request.tools.clone(),
    )
    .map_err(|error| OneStepError(error.to_string()))?;
    let final_response =
        complete(&followup, config, transport).map_err(|error| OneStepError(error.to_string()))?;
    if !final_response.tool_uses().is_empty() {
        return Err(OneStepError(
            "second model call requested another tool; Agent Loop is M3".into(),
        ));
    }
    if final_response.text().is_empty() {
        return Err(OneStepError(
            "second model call returned no final text".into(),
        ));
    }
    Ok(OneStepResult {
        first_response: first,
        tool_result: result,
        followup_request: followup,
        final_response,
    })
}
